notify/notify.go

```diff
@@ -119,109 +143,271 @@ func Now(ctx context.Context) (time.Time, bool) {
 	return v, ok
 }
 
-// A Notifier is a type which notifies about alerts under constraints of the
-// given context.
-type Notifier interface {
-	Notify(context.Context, ...*types.Alert) error
+// A Stage processes alerts constraints under the constraints of the given
+// context.
+type Stage interface {
+	Exec(ctx context.Context, alerts ...*types.Alert) ([]*types.Alert, error)
+}
+
+// Builds a RoutingStage as a map of receivers to Stages
+func BuildRoutingStage(confs []*config.Receiver, tmpl *template.Template, meshWait func() time.Duration, inhibitor *inhibit.Inhibitor, silences *meshprov.Silences, ni *meshprov.NotificationInfos, marker types.Marker) RoutingStage {
```

Comment:
This signature is quite long. That's okay for now but the style pattern for long arguments would be:

```go
+func BuildRoutingStage(
    confs []*config.Receiver,
    tmpl *template.Template,
    meshWait func() time.Duration,
    inhibitor *inhibit.Inhibitor,
    silences *meshprov.Silences,
    ni *meshprov.NotificationInfos,
    marker types.Marker,
) RoutingStage {
```